networkx/algorithms/assortativity/correlation.py

```diff
@@ -154,8 +160,16 @@ def degree_pearson_correlation_coefficient(G, x="out", y="in", weight=None, node
     """
     import scipy as sp
 
+    if nx.is_empty(G):
+        raise nx.NetworkXError(
+            "Cannot compute correlation coefficient on an empty graph."
+        )
```

Comment:
Assuming my reasoning in the above comment is correct (please double-check!), I think the real condition that we want to catch here is that the graph *with self-loop edges removed* is not empty. One way to check that is by comparing the edge set to the set of self-loop edges, e.g. `set(G.edges) - set(nx.selfloop_edges(G))`.

However, this will also lead to subtle changes in the algorithm for the case where the graph with multiple nodes that *only* contain self-loop edges. Right now, this case gives:

```python
>>> G = nx.Graph([(1, 1), (2, 2), (3, 3)])
>>> nx.degree_pearson_correlation_coefficient(G)
ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
nan
```

whereas after the change this would raise an exception instead of producing the `nan`. This seems reasonable to me, but I'm no expert!